keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+   in `foo-ns-1` and `foo-ns-2` and the other with pods in `bar-ns-1`, where inter-tenant
+   traffic is denied.
+
+3. As a cluster administrator, I want to optionally also deploy an additional default
+   set of policies to all workloads that may be overridden by the developers if needed
+
+   Example: I would like to explicitly delegate the restriction of traffic destined
+   for cluster monitoring pods to the developer, allowing them to setup network policy
+   to deny or allow the traffic from/to their application.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Interoperability with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
+a cluster. Cluster scoped controls for north-south traffic may be addressed via
+future versions of the api resources introduced in this or other future KEPs.
+For the time being, the AdminNetworkPolicy resource introduced by this KEP will
+never affect north-south traffic, and thus also don't override or bypass NetworkPolicies
+with ipBlock rules that select external traffic.
+
+## Proposal
+
+In order to achieve the three primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `policy.networking.k8s.io` API group:
+- AdminNetworkPolicy
+
+### AdminNetworkPolicy resource
+
+An AdminNetworkPolicy (ANP) resource will help the administrators:
+1. Set strict security rules for the cluster, i.e. a developer CANNOT override
+these rules by creating NetworkPolicies that applies to the same workloads as the
+AdminNetworkPolicy does.
+2. Set baseline security rules that describes default connectivity for cluster
+workloads, which CAN be overridden by developer NetworkPolicies if needed.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, AdminNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. AdminNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the AdminNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all further rules from all 
+  ANPs and instead be enforced by the K8s NetworkPolicies. If there is no K8s 
+  NetworkPolicy rule match, traffic will be allowed.
```

Comment:
If I understand, the `Pass` action bypasses all ANPs, including "baseline" ANPs.

I agree that's really confusing though and probably shouldn't be the way this works.

Reply:
Maybe this is crazy, and I don't really like it as an idea to be honest, but would it help to make this all fit together better if we reserved one priority for namespaced NetworkPolicies?

e.g., 1-499 is ANP, 500 is NP, and 501-1000 is baseline ANPs that come after? Or 0 is baseline, 1 is NP, and 2-999 is ANP? you get the idea.

It's kinda gross but at the same time I feel like it's going to be hard to explain the semantics of this if NPs are magically inserted at priority "0.5"

Reply:
what is the rationale behind `Pass` ? if there are no rules NetworkPolicies has to prevent too, right?

Reply:
> If I understand, the Pass action bypasses all ANPs, including "baseline" ANPs. 

Sorry these bits need to be explicitly defined in some more detail, ^ was not the intended design, if traffic hit's a `Pass` rule it will skip remaining _Positive_ priority CNP rules, be enforced by Network Policy, and then be enforced by any CNP rules with priority "0" 

The Priority "0" ANP's are meant to allow the Admin to "flip" the default security stance of the cluster from default allow to default deny if desired

Reply:
I had suggested earlier that some use cases might require/strongly-desire multiple priority levels of baseline ANPs for all the same reasons you want multiple priority levels of non-baseline ANPs. (Especially: it's just easier to express some combinations of desired policies that way.)

I'd suggest positive priorities are "normal" ANPs, 0 is NP (and not allowed as an ANP priority), and negative priorities are "baseline" ANPs.

Reply:
Interesting semantic question - are NPs "inserted" at 0  (e.g. if you find no other ANPs you "fall thru" to NP, or is only a "Pass" that activates NP ?

Reply:
In one sketch, I suggested that maybe "Pass" is actually a call to the (possibly-terminal, possibly-not) NetPol rules.  IOW, if a flow doesn't match anything in (namespaced) NetPol then it just "returns" back to ANP and resumes processing rules of lower priority.

To *me* that is easier to understand.

Reply:
I think that's more of a "Jump" action rather than "Pass", which I think have slightly different scenarios in which they'd be used.

Reply:
> I had suggested earlier that some use cases might require/strongly-desire multiple priority levels of baseline ANPs for all the same reasons you want multiple priority levels of non-baseline ANPs. 

So in this case we only have a single user story that references priority="0" rules (`Story 5: Cluster Wide Default Guardrails`) , and that is to flip the default "stance" of the cluster from allow to deny.  Therefore, we didn't see the need to provide multiple "baseline" priorities.  Even so the multiple "appliedTos" allows for some flexibility at the `priority=0` level 

> Interesting semantic question - are NPs "inserted" at 0 (e.g. if you find no other ANPs you "fall thru" to NP, or is only a "Pass" that activates NP ?

NPs are inserted at 0.5 (lets call it that instead of 0 like Casey did earlier), Your traffic can hit netpols by either falling though ANP's or being explicitly pushed there with a matching `Pass` action. 

> if a flow doesn't match anything in (namespaced) NetPol then it just "returns" back to ANP and resumes processing rules of lower priority.

To Me this is confusing :), also from an implementation POV it seems like it would be pretty difficult to implement (thinking varying terminal locations + JUMPTO commands) vs a simple "If you hit `Pass` go to Netpol" sort of directive